src/azstoragetorch/io.py

```diff
@@ -153,8 +158,58 @@ def _blob_url_has_sas_token(self, blob_url: str) -> bool:
         # key to determine if the URL has a SAS token.
         return "sig" in parsed_qs
 
+    def _readline(self, size: Optional[int]) -> bytes:
+        consumed = b""
+        if size == 0 or self._is_at_end_of_blob():
+            return consumed
+
+        limit = self._get_limit(size)
+        if self._readline_buffer:
+            consumed = self._consume_from_readline_buffer(consumed, limit)
+        while self._should_download_more_for_readline(consumed, limit):
+            self._readline_buffer = self._client.download(
+                offset=self._position, length=self._READLINE_PREFETCH_SIZE
+            )
+            consumed = self._consume_from_readline_buffer(consumed, limit)
+        return consumed
+
+    def _get_limit(self, size: Optional[int]) -> Optional[int]:
+        limit = size
+        if size is not None and size < 0:
+            limit = None
+        return limit
+
+    def _consume_from_readline_buffer(
+        self, consumed: bytes, limit: Optional[int]
+    ) -> bytes:
+        if limit is not None:
+            limit -= len(consumed)
+        find_pos = self._readline_buffer.find(self._READLINE_TERMINATOR, 0, limit)
+        end = find_pos + 1
+        if find_pos == -1:
+            buffer_length = len(self._readline_buffer)
+            if limit is None:
+                end = buffer_length
+            else:
+                end = min(buffer_length, limit)
+        consumed += self._readline_buffer[:end]
+        self._readline_buffer = self._readline_buffer[end:]
```

Comment:
Not sure how much we care here but `_readline_buffer`is a `bytes` and so slicing it does a copy right? Would it be worth using a `bytearray` to avoid the copies?

Reply:
I'm not sure if a `bytearray` will help much here mainly it is just a mutable form of `bytes`. It still returns a copy of the bytes when sliced in order to create a new `bytearray` object. I think exploring using [`memoryview`](https://docs.python.org/3/library/stdtypes.html#memory-views) objects could be interesting since those do not copy on splices. I'll add a note about considering converting it to a `memoryview`. I've been looking more into `memoryview`s as part of the write implementation so we can circle back to this later.

Reply:
Talked more offline about this... The `bytearray` would help when we concatenate the readline buffer slice to `consumed` as there are performance benefits because the `consumed` bytearray would be mutable. That being said, we are not doing these computations in a tight loop; the download itself will take much more time. So any potential performance gain will be minimal if not negligible.